### ISSUE TITLE
Extend type extension mechanism to support generic types

### ISSUE THREAD

**Issue by En3Tho** [fsharp/fslang-suggestions#1082]
# Title of Suggestion

I propose we improve type extesions mechanism to support generic types extension eg:

``` fsharp
type ('a when 'a :> ICollection<'b>) with
        member inline this.Yield (value: 'b) = this.Add value
        member inline this.YieldFrom (values: 'b seq) = for value in values do this.Add value
        member inline this.While (moveNext, whileExpr: unit -> unit) = while moveNext() do whileExpr()
        member inline this.For (values, forExpr: 'b -> unit) = for value in values do forExpr value
        member inline this.Combine (_, cexpr) = cexpr()
        member inline this.Zero() = ()
        member inline this.Delay beforeAdd = beforeAdd
        member inline this.Run (runExpr: unit -> unit) = runExpr(); this
```

or 
``` fsharp
type SCollection<'collection, 'item when 'collection :> ICollection<'item>> = 'collection
type SCollection<'collection, 'item> with
        member inline this.Yield (value:  'item) = this.Add value
        member inline this.YieldFrom (values:  'item seq) = for value in values do this.Add value
        member inline this.While (moveNext, whileExpr: unit -> unit) = while moveNext() do whileExpr()
        member inline this.For (values, forExpr:  'item -> unit) = for value in values do forExpr value
        member inline this.Combine (_, cexpr) = cexpr()
        member inline this.Zero() = ()
        member inline this.Delay beforeAdd = beforeAdd
        member inline this.Run (runExpr: unit -> unit) = runExpr(); this
```

The existing way of approaching this problem in F# is either to make CE for every possible type to save the type itself or extend in interface and lose the type (but upcast it later):

``` fsharp
type ICollection<'a> with
        member inline this.Yield (value: 'a) = this.Add value
        member inline this.YieldFrom (values: 'a seq) = for value in values do this.Add value
        member inline this.While (moveNext, whileExpr: unit -> unit) = while moveNext() do whileExpr()
        member inline this.For (values, forExpr: 'a -> unit) = for value in values do forExpr value
        member inline this.Combine (_, cexpr) = cexpr()
        member inline this.Zero() = ()
        member inline this.Delay beforeAdd = beforeAdd
        member inline this.Run (runExpr: unit -> unit) = runExpr(); this

let items = ResizeArray() { 1; 2; 3 } // ICollection<int>, not ResizeArray<int>
```

## Pros and Cons 

The advantages of making this adjustment to F# are:
1. The ability to make a generic base type / interface extensions with saving the original type
2. Potentially impoving performance if people are using interface based CE's  (eg. no boxing on structs but I'm not sure people use them in CE's)

The disadvantages of making this adjustment to F# are:

I personally don't see any. I believe this improvement is not harder to understand or easier to abuse than existing capabilities of type extension.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL):  L/XL (I believe it might be hard to do this but maybe I'm wrong). I believe it should be possible to make becuase in the end of the day it's just functions with a certain shape and compiler hint.

Related suggestions: (put links to related suggestions here)

## Affidavit (please submit!)

I'm not sure what to tick here because I've failed to find in issue like this but I believe it's been around for long time already.

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [ ] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [ ] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by gsomix** [comment]
@En3Tho Surprisingly, it's possible now using ["C# style extension members"](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/type-extensions#extension-methods).

```fsharp
open System.Collections.Generic
open System.Runtime.CompilerServices
 
[<Extension>]
type ICollectionExtensions() =
    [<Extension>]
    static member Yield(this: 'a when 'a :> ICollection<'b>, value: 'b) = this.Add value
    
    [<Extension>]
    static member YieldFrom(this: 'a when 'a :> ICollection<'b>, values: 'b seq) = for value in values do this.Add value
    
    [<Extension>]
    static member While(this: 'a when 'a :> ICollection<'b>, moveNext, whileExpr: unit -> unit) = while moveNext() do whileExpr()
    
    [<Extension>]
    static member For(this: 'a when 'a :> ICollection<'b>, values, forExpr: 'b -> unit) = for value in values do forExpr value
    
    [<Extension>]
    static member Combine (this: 'a when 'a :> ICollection<'b>, _, cexpr) = cexpr()
    
    [<Extension>]
    static member Zero(this: 'a when 'a :> ICollection<'b>) = ()
    
    [<Extension>]
    static member Delay(this: 'a when 'a :> ICollection<'b>, beforeAdd) = beforeAdd
    
    [<Extension>]
    static member Run(this: 'a when 'a :> ICollection<'b>, runExpr: unit -> unit) = runExpr(); this
```

Usage:

```fsharp
let xs = ResizeArray() { for i in 1..10 do yield i*i }
val xs : ResizeArray<int> // seq [1; 4; 9; 16; ...]
```

**Comment by En3Tho** [comment]
@gsomix Thanks! Didn't know this.
Do you think there is a need for single unified syntax?
Basically this solves my problem and it's not hard to write few more lines of code but I'm still interested in opinion of community.

**Comment by dsyme** [comment]
> Do you think there is a need for single unified syntax?

Please add an issue to track this.

There are two problems

1. Unifying the F# syntax.  

2. Unifying the compiled representations with C#.  

We didn't unify in order to ensure F#-authored "C# style extension methods" could be used from C#.  We couldunify but this really requires C# to have "extension everything" to match the F# extension properties.  See https://github.com/dotnet/csharplang/issues/192

Anyway, ideally we could unify the F# syntax, and I think it would make things simpler for the F# programmer to just use that if they don't care about interop

**Comment by En3Tho** [comment]
@dsyme while trying this thing out I found that "C# style extension method"-builders break existing ones - like async.
The code like this 
``` fsharp
module ICollectionBuilder =

    [<AbstractClass;Extension>]
    type ICollectionExtensions() = // TODO: F# 6 InlineIfLambda
        [<Extension>]
        static member inline Yield(collection: #ICollection<'a>, value: 'a) = collection.Add value
        [<Extension>]
        static member inline YieldFrom(collection: #ICollection<'a>, values: 'a seq) = for value in values do  collection.Add value
        [<Extension>]
        static member inline Run(collection: #ICollection<'a>, runExpr: RunExpression) = runExpr(); collection
```
breaks code like this:
``` fsharp
[<Fact>]
let ``Test that async computation expression compile`` () =
    let rec taskTest job = async { // <- The type 'AsyncBuilder' is not compatible with the type 'System.Collections.Generic.ICollection<'a>'
        return! job
    }
    let expected = 0
    let result = taskTest (Async.ofObj expected) |> Async.RunSynchronously
    Assert.Equal(result, expected)
```
I believe it has to do with overload resolution priority. If my memory is right, AsyncBuilder is made with different priority extensions too and seems like Compiler is unable to throw the #ICollection<'a> constraint out of the candidate list. I've heard about such behaviour some time ago - like compiler is picking first match, not the best one.
Can this be somehow fixed or is it a huge breaking change?

**Comment by dsyme** [comment]
@En3Tho FIrst, you can omit the `#` though I haven't checked if it makes a difference:

```fsharp
        static member inline Yield(collection: ICollection<'a>, value: 'a) = collection.Add value
```

> I believe it has to do with overload resolution priority. If my memory is right, AsyncBuilder is made with different priority extensions too ...

Intrinsic methods always take precedence of extension methods - this is a C# rule used in F# as well.  I expect that's what you're hitting here. Overall this makes extending builders with new overloads really problematic and not something that's to be recommended unless the builder has been designed and tested with that scenario in mind.

**Comment by En3Tho** [comment]
@dsyme The problem is that I need "Run" method to be # to save the type itself instead of returning an interface.
I know about precedence but I personally find it strange that compiler is arguing about overloads when their constrains are obviously not satisfied. Like AsyncBuilder is obviously not an ICollection, not an IDictionary, not an "Addeable" type, etc.

And the point is that I do not want to extend AsyncBuilder, I want to extend another completely different thing. But async builder is the one to break for some reason...

**Comment by gsomix** [comment]
@En3Tho There is no `Run` method for `AsyncBuilder`. And it might be reason why compilation of CE fails somehow.

**Comment by En3Tho** [comment]
@gsomix Yeap. It is done via different priority extensions and compiler somehow "forgets" about them. F# 6's Task builder is done like this too...

**Comment by gsomix** [comment]
@dsyme @En3Tho It seems that `Run` extension method shouldn't have resolved for `AsyncBuilder`. Is [this bug](https://github.com/dotnet/fsharp/issues/11658) related to the problem?

**Comment by En3Tho** [comment]
I guess it is. Seems awfully similar.

**Comment by dsyme** [comment]
@En3Tho Why are you trying to use a collection type as a builder type?  That's an unusual approach - I'm not saying it can't be done, but it seems unlikely to work well.   Certainly adding builder methods like `Run` generically looks unlikely to work as well.

> @dsyme @En3Tho It seems that Run extension method shouldn't have resolved for AsyncBuilder. 

No, it's by specification - if there is a Run method in scope - including generic extension methods that add a Run to an arbitrary type `T` (even if constrained) - then it will be used. The method applicability is determined later. That is https://github.com/dotnet/fsharp/issues/11658 is considered as by design as things stand - the method is added to the available overloads and overload resolution resolves things.

In short extending an arbitrary `T` by `Run` is not going to work well (even if T is constrained).

**Comment by En3Tho** [comment]
@dysme
The idea is to use collections same as existring builders and F#: seq, [], [||] etc.
Out of the box ResizeArray, HashSet, Stack, Queue whatever are not supported. You can add StringBuilder here too.

``` fsharp
// Why use seq or [| |] if you need a ResizeArray? Use ResizeArray instead.
ResizeArray() { 1; 2; 3; yield! values }
HashSet() { v1; v2; yield! values }
```

``` fsharp
StringBuilder() { // if done via srtp compiler will resolve types automatically with only single function
    1
    2.
    3.f
    '4'
    "5"
}
```
Recent example with gui.cs's View class (it has Add method to add a child view):
![image](https://user-images.githubusercontent.com/37334640/135616558-eee5f69c-2a4a-470a-a6a4-8e919d36b789.png)
became moch more readable:
``` fsharp
Application.Top {
        window {
            keyQueryFrameView {
                keyQueryTextField
            }
            keyQueryFilterFrameView {
                keyQueryFilterTextField
            }
            dbPickerFrameView
            dbPickerComboBox
            keysFrameView {
                keysListView
            }
            resultsFrameView {
                resultsListView
            }
            commandFrameView {
                commandTextField
            }
            resultFilterFrameView {
                resultFilterTextField
            }
        }
    } |> ignore
```

All of these types are "Builders" by nature, eg ICollection builds a set of values, StringBuilder builds a String, Gui.cs's View builds a set of child Views.

F# gives a good opportunity to make code simplier and easier when using them.

**Comment by Happypig375** [comment]
We can add this to FSharp.Core too, and we will have an equivalent to C#'s collection initializers.

**Comment by En3Tho** [comment]
@Happypig375 Yes that's basically the idea.

**Comment by dsyme** [comment]
I see, thanks

Perhaps better would be
```
ResizeArray { 1; 2; 3; yield! values }
HashSet { v1; v2; yield! values }
```

This may need to be a first class language feature

**Comment by En3Tho** [comment]
@dsyme I believe constructors are better because you can explicitly preallocate internal storages.

**Comment by dsyme** [comment]
> @dsyme I believe constructors are better because you can explicitly preallocate internal storages.

Yes I guess you could need construction parameters too
```fsharp
Thing(200) { 1; 2; 3; yield! values }
```

> My idea was to make this feature via extensions which I think is logical from F#'s point of view. But maybe I'm mistaken.

It is logical, yes, though when designing CEs we always assumed a separate nominal builder type.  There are a lot of ramifications to allowing concrete data types as the builder that we'd need to consider.

**Comment by En3Tho** [comment]
I see. For now I resort to inline to make speed comparable. The downside is that naive for expression (for i in 0 to n do i) can't be inlined because it turns to seq. So when for expressions like this are used, hand written version is faster.
 InlineIfLambda attribute helps a little bit - asm code for simple builder with for, while and try with expressions shrinked by 15 bytes compared to naive inline version. Not much but still good. Maybe the gains will be bigger in a complex case.
 When both versions use actual seq with for expression then allocations are the same. Perf is still on the side of hand written version (about 10% faster) and I believe this is a limitation of JIT's locals tracking feature. But this needs more investigation.

**Comment by uxsoft** [comment]
By the way, if your goal is to build UI using computation expressions, I have a few projects looking at this problem. Maybe you'll find some inspiration: 
- [https://github.com/uxsoft/Fable.AntDesign](https://github.com/uxsoft/Fable.AntDesign)
- [https://github.com/uxsoft/FuncUI.Experiments](https://github.com/uxsoft/FuncUI.Experiments)
- [https://github.com/uxsoft/FUI](https://github.com/uxsoft/FUI)

**Comment by Happypig375** [comment]
C# collections already use `AddRange` as the `AddMany` of F#. How to guarantee interop here?

**Comment by En3Tho** [comment]
Found more cases when "generic" ExtensionMethod break existing builders. Now with option builder by @dsyme (https://github.com/dotnet/fsharp/blob/main/tests/benchmarks/TaskPerf/option.fs)
Adding an extension like this
``` fsharp
[<Extension; EditorBrowsable(EditorBrowsableState.Value)>]
    static member inline While(_, [<InlineIfLambda>] moveNext: unit -> bool, [<InlineIfLambda>] whileExpr: unit -> unit) : CollectionCode =
        fun () -> while moveNext() do (whileExpr())
```
Completely breaks option builder's own While method. I thought the case was only when builder type has no own method so extension takes precedence, but seems like this is not the case?

**Comment by gsomix** [comment]
@En3Tho That's because `while` loop compiles with `Zero` (if available) when a computation does not return a value.

**Comment by En3Tho** [comment]
Hm. I guess I've added it myself. It is defined like this:
``` fsharp
type OptionBuilderUsingInlineIfLambdaBase() =
    member inline _.Zero _ : OptionCode<unit> =
        fun() -> ValueSome ()
   ...
```

**Comment by gsomix** [comment]
@En3Tho Btw shouldn't it return `fun () -> ValueNone` instead?

**Comment by En3Tho** [comment]
@gsomix It exists the while loop on ValueNone so ValueSome() is needed.

**Comment by gsomix** [comment]
@En3Tho Ah, I see. Thanks!

I've tried with a little bit different builder using explicit `Zero` call from `While` method. I agree, that doesn't look right. Please note type of `zero` let-binding.

```fsharp
open System.Runtime.CompilerServices

type OptionBuilder() =
    member inline _.Return(x) = Some x

    member inline _.Bind(m : Option<'a>, f: 'a -> Option<'b>) = Option.bind f m

    member inline this.Zero _ = this.Return ()

    member inline _.Delay(f: unit -> _) = f

    member inline _.Run(f) = f()

    member this.While
        (guard: unit -> bool, generator: unit -> Option<_>)
        : Option<_> =
        if not (guard ()) then 
            this.Zero ()
        else 
            this.Bind(this.Run generator, fun () -> this.While (guard, generator))
               
let option = OptionBuilder()

[<Extension>]
type Ext =
    [<Extension>]
    static member inline Zero _ = fun() -> ()

let zero = option.Zero // val foo: 'a -> unit option

let foobar() = 
    option {
        while true do
            printfn "LOL" 
            // error FS0001: This expression was expected to have type
            //     'Option<unit>'    
            // but here has type
            //     'unit -> unit'
    }
```

**Comment by gsomix** [comment]
@dsyme @En3Tho It seems something wrong with overload resolving. Please compare identical code in C# and F#.

F# code, [link to sharplab](https://sharplab.io/#v2:DYLgZgzgNALiBOBXAdgHwPYAcCmyAEAygJ4QzYC2AsAFBa6ElnkB0ASijAJbnbMDC6cpk7Bs8AmIBunAMbYINGjCI48AMXToARgEN4ACgCUeALw08FvD3JaxeAB7MAsvr0BzEHgDkAVWMm8AWQIdFFmAHV4TjIAGU5kbH0AIidsGAALdAATJMNzSxoAbQAeAFF7MmDOdGQAPgBdJRVsPHKYU3yLErbcCGq6xupLPFIdLhkrClt4PBd40eQ5Ty8AFSg8d094mH9AmpCwyOjsOITknqqayYzs3M68GlF2sE1dGYCNbT0jGhev+Gc+gALAAmQxAA===)
```fsharp
open System
open System.Runtime.CompilerServices

type Foobar() =
    member x.M(arg: 'U) = Console.WriteLine("Method")
    
[<Extension>]
type Ext =
    [<Extension>]
    static member M(instance: 'T, arg: int) = Console.WriteLine("Extension method")
    
let foobar = Foobar()
foobar.M(42) // `M` call optimised away, prints "Extension method"
```

and C# code, [link to sharplab](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQcDMACR2BiA9oQEYCGY2A3utnbjnACzYCyAPAKoB8AFJ9goBzAJTVa9SXCQBOXgCJWAU2AALQgBN5IgNwS6AX3RG06LLiQA2XAmwBRAB7B0NNFMZXcLDgBU+agEsAZ2wfbACoIOAyKABjJQAacKhgQTBRfXE3SXppOXlHYCVIgMIobABbFXUtXUyTEzMETBdMxARM1xz6ADcKbAAzYnJKAF5sKCUAdwJhil467O6h0goAOlZeJgRFyQagA=)
```csharp
using System;
public class Foobar {
    public void M<U>(U arg) {
        Console.WriteLine("Method");
    }
}

public static class Ext
{
    public static void M<T>(this T instance, int arg)
    {
        Console.WriteLine("Extension method");
    }
}

public static class Program
{
    public static void Main()
    {
        var foobar = new Foobar();
        foobar.M(42); // prints "Method"
    }
}
```

**Comment by En3Tho** [comment]
@gsomix Wow. Thanks. This doesn't really seem to be by design anymore...

**Comment by gsomix** [comment]
@En3Tho Yeah, example with `Zero` is even harder because it's "concrete method with generic arg" vs "extension method on generic type without args (unit?)".

But for "C# style extension members" I'd prefer same behaviour as for C#.

**Comment by gsomix** [comment]
@En3Tho So rule of thumb here — don't use wildcards for know arguments. E.g. for `Zero: unit -> smth`. But agree it might be confusing. :)

**Comment by En3Tho** [comment]
@gsomix Even if method is defined like 
``` fsharp
[<Extension;EditorBrowsable(EditorBrowsableState.Value)>]
    static member inline Zero(_, _: unit) : CollectionCode = fun() -> ()
```
still breaks other stuff :(
I believe "proper" signature is not the main problem here

**Comment by gsomix** [comment]
@En3Tho Please define original `Zero` method as

```fsharp
type OptionBuilderUsingInlineIfLambdaBase() =
    member inline _.Zero(): OptionCode<unit> =
        fun() -> ValueSome ()
        (* ... *)
```

**Comment by En3Tho** [comment]
It works this way but breaks otherwise. Funny.

**Comment by dsyme** [comment]
See also https://github.com/fsharp/fslang-suggestions/issues/858

**Comment by dsyme** [comment]
I'm going to close this as effectively tracked by https://github.com/fsharp/fslang-suggestions/issues/858 - or at least if we do one of these we would do both
